src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs

```diff
@@ -244,7 +244,15 @@ private Task<CollectionResult> BindComplexCollection(ModelBindingContext binding
             Logger.AttemptingToBindCollectionUsingIndices(bindingContext);
 
             var indexPropertyName = ModelNames.CreatePropertyModelName(bindingContext.ModelName, "index");
-            var valueProviderResultIndex = bindingContext.ValueProvider.GetValue(indexPropertyName);
+
+            // Remove any value provider that may not use indexPropertyName as-is. Don't match e.g. Model[index].
+            var valueProvider = bindingContext.ValueProvider;
+            if (valueProvider is IKeyRewriterValueProvider keyRewriterValueProvider)
+            {
+                valueProvider = keyRewriterValueProvider.Filter() ?? new CompositeValueProvider();
```

Comment:
Is the '??' here just to handle the possibility that we're operating on something other than the composite value provider?

Reply:
It's here for a few reasons
- Something other than `CompositeValueProvider` might return `null`
- At present at least, `CompositeValueProvider.Filter()` will return `null` in some cases
- It matches the `IBindingSourceValueProvider` usage pattern e.g. https://github.com/aspnet/Mvc/blob/58026eacbd6d2411fb49f9ca5858617d9641f5b5/src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs#L323